File: 78-subsets/subsets.cpp
class Solution {
public:
    vector<vector<int>> res; 
    void Solve(vector<int>& nums , int i , int n,vector<int> &temp){
        if(i >= n){
            res.push_back(temp);
            return;
        }

        temp.push_back(nums[i]);
        Solve(nums, i+1 , n , temp);
        temp.pop_back();
        Solve(nums , i + 1 , n, temp);
    }
    vector<vector<int>> subsets(vector<int>& nums) {
        int n = nums.size();
        vector<int> temp;
        Solve(nums ,0 , n , temp );
        
        return res;
    }
};

File: 78-subsets/subsets.java
class Solution {
    List<List<Integer>> res = new ArrayList<>();
    private void Solve(int [] nums, int n , int i , List<Integer> temp ){
        if( i>= n){
            res.add(new ArrayList<>(temp));
            return;
        }

        temp.add(nums[i]);
        Solve(nums , n , i+1 , temp);
        temp.remove(temp.size() - 1);
        Solve(nums , n , i+1 , temp);
    }

    public List<List<Integer>> subsets(int[] nums) {
        int n = nums.length;
        List <Integer> temp = new ArrayList<>();
        Solve(nums, n , 0 , temp);

        return res;
    }
}